File: 2020/07/haverstacks.rs
use std::{
    collections::HashMap,
    io::{self, BufRead, Error, ErrorKind},
};

type Color = String;
type Inside = (Color, u8);
type Bags = HashMap<Color, Vec<Inside>>;

fn main() -> io::Result<()> {
    let stdin = io::stdin();

    let mut bags: Bags = HashMap::new();
    for line in stdin.lock().lines() {
        match parse_line(line?) {
            Ok((bag, in_bag)) => {
                bags.insert(bag, in_bag);
            }
            Err(err) => return Err(Error::new(
                ErrorKind::InvalidInput,
                format!("While parsing stdin: {}", err)
            ))
        }
    }

    let mut cache = HashMap::new();
    let mut nb = 0;
    for bag in bags.keys() {
        if contains_gold(bag, &bags, &mut cache) {
            nb += 1;
        }
    }

    println!("{}\n{}", nb, bags_in(&String::from("shiny gold"), &bags));

    return Ok(());
}

fn bags_in(color: &Color, bags: &Bags) -> u64 {
    return bags.get(color).unwrap()
        .iter()
        .map(|x| x.1 as u64 * (bags_in(&x.0, bags) + 1))
        .sum();
}

fn contains_gold(color: &Color, bags: &Bags, cache: &mut HashMap<Color, bool>) -> bool {
    if let Some(b) = cache.get(color) {
        return *b;
    }

    let inside = bags.get(color).unwrap();
    if let Some(_) = inside.iter().find(|x| x.0 == "shiny gold") {
        cache.insert(color.clone(), true);
        return true;
    } else {
        for bag in inside {
            if contains_gold(&bag.0, bags, cache) {
                return true;
            }
        }

        cache.insert(color.clone(), false);
        return false;
    }
}

fn get_two<'a, I>(words: &mut I) -> Option<String>
where
    I: Iterator<Item = &'a str>
{
    return words.next()
        .and_then(|x| words.next()
                  .map(|y| x.to_owned() + " " + y));
}

fn parse_line(line: String) -> Result<(Color, Vec<Inside>), String> {
    let mut words = line.split(' ');
    let mut res = match get_two(&mut words) {
        Some(s) => (s, vec![]),
        None => return Err(String::from("Unexpected end of line"))
    };

    match words.next() {
        Some("bags") => {}
        Some(s) => return Err(format!("Unexpected word: {}", s)),
        None => return Err(String::from("Unexpected end of line"))
    }

    match words.next() {
        Some("contain") => {}
        Some(s) => return Err(format!("Unexpected word: {}", s)),
        None => return Err(String::from("Unexpected end of line"))
    }

    loop {
        match words.next() {
            Some("no") => break,
            Some(num) => match num.parse::<u8>() {
                Ok(n) => if let Some(s) = get_two(&mut words) {
                    res.1.push((s, n));
                } else {
                    return Err(String::from("Unexpected end of line"));
                }
                Err(_) => return Err(format!("Wanted digit, got {}", num))
            }
            None => break
        }

        match words.next() {
            Some("bag,") => {}
            Some("bags,") => {}
            Some("bag.") => break,
            Some("bags.") => break,
            Some(s) => return Err(format!("Unexpected word: {}", s)),
            None => return Err(String::from("Unexpected end of line"))
        }
    }

    return Ok(res);
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parsing() {
        assert!(parse_line(String::from(
            "light red bags contain 1 bright white bag, 2 muted yellow bags."
        )) == Ok((String::from("light red"), vec![
            (String::from("bright white"), 1),
            (String::from("muted yellow"), 2),
        ])));

        assert!(parse_line(String::from(
            "dark orange bags contain 3 bright white bags, 4 muted yellow bags."
        )) == Ok((String::from("dark orange"), vec![
            (String::from("bright white"), 3),
            (String::from("muted yellow"), 4),
        ])));

        assert!(parse_line(String::from(
            "faded blue bags contain no other bags."
        )) == Ok((String::from("faded blue"), vec![])));
    }
}

File: 2020/07/haverstacks.go
package main

import (
	"fmt"
	"os"
	"bufio"
	"strconv"
	"strings"
	"errors"
)

type inside []struct {
	number int
	bag string
}

type bags map[string]inside

func readLine(s string) (color string, in inside, err error) {
	words := strings.Split(s, " ")
	if len(words) < 7 {
		err = errors.New("readLine: Line too short")
		return
	}

	color = strings.Join(words[0:2], " ")

	if words[4] == "no" {
		return
	}

	nbInside := len(words) / 4 - 1
	in = make(inside, nbInside)

	for i := 0; i < nbInside; i++ {
		place := (i + 1) * 4
		nb, e := strconv.Atoi(words[place])
		if e != nil {
			err = e
			return
		}
		in[i].number = nb
		in[i].bag = strings.Join(words[place + 1 : place + 3], " ")
	}

	return
}

func containsGold(color string, bs bags, cache map[string]bool) bool {
	if res, exists := cache[color]; exists {
		return res
	}

	for _, v := range bs[color] {
		if v.bag == "shiny gold" {
			return true
		}

		c := containsGold(v.bag, bs, cache)
		cache[v.bag] = c

		if c {
			return true
		}
	}

	return false
}

func bagsIn(color string, nb int, bs bags, ch chan<- int) {
	c := make(chan int)
	in := bs[color]
	for _, v := range in {
		go bagsIn(v.bag, v.number, bs, c)
	}

	sum := 0
	for i := 0; i < len(in); i++ {
		sum += <-c
	}

	ch <- (sum + 1) * nb
}

func main() {
	scanner := bufio.NewScanner(os.Stdin)

	bags := make(bags)
	for scanner.Scan() {
		color, in, err := readLine(scanner.Text())
		if err != nil {
			fmt.Fprintln(os.Stderr, "Error while parsing stdin:", err)
			os.Exit(1)
		}

		bags[color] = in
	}

	if err := scanner.Err(); err != nil {
		fmt.Fprintln(os.Stderr, err)
		os.Exit(1)
	}

	cache := make(map[string]bool)
	sum := 0
	for color := range bags {
		if containsGold(color, bags, cache) {
			sum++
		}
	}

	fmt.Println(sum)

	ch := make(chan int)
	go bagsIn("shiny gold", 1, bags, ch)

	fmt.Println(<-ch - 1)
}
